rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,110 @@
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Alejandro Guillen <alejandro@iasql.com>
+
+## Summary
+
+Adding new features to get closer to the goal of the EC2 module's completeness adds more complexity to the model. This complexity directly affects the user experience, especially for new users, since they can feel overwhelmed if they encounter a complicated model.
+
+The definition of a solid base now will imply that new features added to this module will not translate into breaking changes to the model.
+
+Taking a look at the AWS EC2 console options, we would still need to add the following to reach EC2 module completeness:
+
+- Instance type - Availability zone relationship
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Saving plans
+- Reserved instances
+- Scheduled instances
+- Capacity reservations
+- Auto Scaling Groups
+- AMIs management
+- EBS volume snapshots and life cycle
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+Almost all the features listed will need a direct integration with the `instance` table, needing to add at least eight more columns to the model. Those columns are only for integration, but the `instance` model still needs more properties to be defined to be complete. Some of those missing properties are: instance auto-recovery, shutdown behaviour, stop - hibernate behaviour, termination protection, stop protection, detailed cloudWatch monitoring, elastic GPU, credit specification, tenancy, RAM disk ID, kernel ID or Metadata accessible (I might be missing others).
+
+The current instance module has ten columns. Adding the ones representing integration with other features will almost double its size.
+
+## Proposal
+
+The `instance` table is the core of the `aws_ec2` module and is the one that can grow more in complexity. The proposal is to keep adding the necessary columns to the `instance` table. It is better to have a unique table to look for instances than increase the complexity of the model, creating other tables and adding to the users the need to understand our model first. Eventually, we can add an `aws_ec2_simplified` module that can expose the basics to create an instance and leave the `aws_ec2` for advanced users.
+
+We can divide the feature list as follow:
+
+The direct implication in the `instance` table:
+
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Capacity reservations
+- Auto Scaling Groups
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+No implicaiton in the `instance` table:
+
+- Instance type - Availability zone relationship
+- AMIs management
+- EBS volume snapshots and life cycle
+
+The `Saving plans`, `Reserved instances` or `Scheduled instances` are not included in the list above since they seem to be budgeting solutions that need interaction with the user rather than infrastructure-related modules' features that could have any implication in the model.
```

Comment:
I don't agree with this. If a user knows the EC2 instance they are going to spin up will be permanently up for years at a time, we should let them specify that so they can pay a significantly lower price for it.

Reply:
Yes, I agree with what you are saying, but the way they work it is at cost/budget level and not resources. Let me give an example. When you purchase a Reserved instance you are not buying an instance for x period of time, you are buying the "spot" (similar when you reserved a spot in a parking lot) for an instance that have those specifications you are defining. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/apply_ri.html

Reply:
Got it, but I would say that could justifiably be a different table in that case, not that we should ignore it entirely.